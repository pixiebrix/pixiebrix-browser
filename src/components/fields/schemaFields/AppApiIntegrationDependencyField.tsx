/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import type React from "react";
import { useField, useFormikContext } from "formik";
import { PIXIEBRIX_INTEGRATION_ID } from "@/services/constants";
import { produce } from "immer";
import { isEqual, set } from "lodash";
import {
  keyToFieldValue,
  type IntegrationsFormSlice,
} from "./integrations/integrationDependencyFieldUtils";
import {
  type Expression,
  type OutputKey,
  type ServiceVarRef,
} from "@/types/runtimeTypes";
import useAsyncEffect from "use-async-effect";

const PIXIEBRIX_OUTPUT_KEY = "pixiebrix" as OutputKey;

/**
 * Schema-based field for the PixieBrix API (@pixiebrix/api).
 *
 * - Does not render a DOM element
 * - Ensures a @pixiebrix/api dependency is included for the extension
 */
const AppApiIntegrationDependencyField: React.FunctionComponent<{
  name: string;
}> = ({ name }) => {
  const { values: root, setValues: setRootValues } =
    useFormikContext<IntegrationsFormSlice>();
  const [
    { value: dependencyOutputKey },
    ,
    { setValue: setDependencyOutputKey },
  ] = useField<Expression<ServiceVarRef>>(name);

  // This currently happens when a brick is copy-pasted into a separate extension
  // that does not yet have root.integrationDependencies configured, but already has the
  // integration dependency key set up in the (copied) BrickConfig.
  const isBadValue =
    dependencyOutputKey &&
    !root.integrationDependencies.some(({ outputKey }) =>
      isEqual(keyToFieldValue(outputKey), dependencyOutputKey)
    );

  useAsyncEffect(
    async () => {
      if (dependencyOutputKey == null) {
        const match = root.integrationDependencies.find(
          ({ integrationId }) => integrationId === PIXIEBRIX_INTEGRATION_ID
        );
        if (match?.outputKey) {
          // If the service is already being used, default to the currently configured auth
          console.debug(
            "PixieBrix dependency already exists for %s, using output key %s",
            match.integrationId,
            match.outputKey,
            { root, match }
          );
          await setDependencyOutputKey(keyToFieldValue(match.outputKey));
        } else {
          console.debug("Adding PixieBrix API dependency");
          await setRootValues(
            produce(root, (draft) => {
              draft.integrationDependencies.push({
                integrationId: PIXIEBRIX_INTEGRATION_ID,
                // XXX: in practice the pixiebrix outputKey won't be used by any other service. However, we might
                // consider using fresh identifier here to eliminate the possibility of colliding with a different
                // service that is somehow already using the @pixiebrix key
                outputKey: PIXIEBRIX_OUTPUT_KEY,
                // PixieBrix service does not use an ID -- the auth is automatically handled based on the logged-in user
                // configId: undefined,
              });

              set(draft, name, keyToFieldValue(PIXIEBRIX_OUTPUT_KEY));
            })
          );
        }
      } else if (isBadValue) {
        // Clearing this "bad value" value will enable the preceding if-branch to
        // execute again, and that will configure root.services properly
        await setDependencyOutputKey(null);
      }
    },
    // Run on mount, or if we detect a "bad value" (see comment above)
    [isBadValue]
  );

  return null;
};

export default AppApiIntegrationDependencyField;
