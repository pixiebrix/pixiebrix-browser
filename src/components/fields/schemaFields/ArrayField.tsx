/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// Ok to use object here since we don't have any key-specific logic
// eslint-disable-next-line @typescript-eslint/ban-types
import React from "react";
import { FieldProps } from "@/components/fields/propTypes";
import { FieldArray, useField } from "formik";
import { Schema } from "@/core";
import { Button, Form } from "react-bootstrap";
import { fieldLabel } from "@/components/fields/fieldUtils";
import { getDefaultField } from "@/components/fields/blockOptions";
import { booleanPredicate, findOneOf, textPredicate } from "./schemaUtils";

function getDefaultArrayItem(schema: Schema): unknown {
  if (schema.default) {
    return schema.default;
  }

  if (textPredicate(schema)) {
    return "";
  }

  if (schema.type === "object") {
    return {};
  }

  if (findOneOf(schema, booleanPredicate)) {
    return false;
  }

  if (findOneOf(schema, textPredicate)) {
    return "";
  }

  return null;
}

const ArrayField: React.FunctionComponent<FieldProps<object[]>> = ({
  schema,
  label,
  ...props
}) => {
  const [field] = useField(props);

  if (Array.isArray(schema.items)) {
    throw new TypeError("Support for arrays of mixed types is not implemented");
  } else if (typeof schema.items === "boolean") {
    throw new TypeError("Schema required for items");
  }

  const schemaItems = schema.items as Schema;

  return (
    <Form.Group controlId={field.name}>
      <Form.Label>{label ?? fieldLabel(field.name)}</Form.Label>
      {schema.description && (
        <Form.Text className="text-muted">{schema.description}</Form.Text>
      )}
      <FieldArray name={field.name}>
        {({ remove, push }) => (
          <>
            <ul className="list-group">
              {(field.value ?? []).map((item: unknown, index: number) => {
                const Renderer = getDefaultField(schemaItems);
                return (
                  <li className="list-group-item" key={index}>
                    <Renderer
                      key={index}
                      name={`${field.name}.${index}`}
                      schema={schemaItems}
                      label={`${label ?? fieldLabel(field.name)} #${index + 1}`}
                    />
                    <Button
                      variant="danger"
                      size="sm"
                      onClick={() => {
                        remove(index);
                      }}
                    >
                      Remove Item
                    </Button>
                  </li>
                );
              })}
            </ul>
            <Button
              onClick={() => {
                push(getDefaultArrayItem(schemaItems));
              }}
            >
              Add Item
            </Button>
          </>
        )}
      </FieldArray>
    </Form.Group>
  );
};

export default ArrayField;
