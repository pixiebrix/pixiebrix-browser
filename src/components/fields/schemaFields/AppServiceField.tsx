/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useEffect } from "react";
import { SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import { useField, useFormikContext } from "formik";
import { Expression, OutputKey, ServiceKeyVar } from "@/core";
import { PIXIEBRIX_SERVICE_ID } from "@/services/constants";
import {
  keyToFieldValue,
  ServiceSlice,
} from "@/components/fields/schemaFields/ServiceField";
import { produce } from "immer";
import { createTypePredicate } from "@/components/fields/fieldUtils";
import { SERVICE_BASE_SCHEMA } from "@/services/serviceUtils";
import { isEqual } from "lodash";

const PIXIEBRIX_OUTPUT_KEY = "pixiebrix" as OutputKey;

export const isAppServiceField = createTypePredicate(
  (schema) => schema.$ref === `${SERVICE_BASE_SCHEMA}${PIXIEBRIX_SERVICE_ID}`
);

/**
 * Schema-based field for the PixieBrix API (@pixiebrix/api).
 *
 * - Does not render a DOM element
 * - Ensures a @pixiebrix/api dependency is included for the extension
 */
const AppServiceField: React.FunctionComponent<SchemaFieldProps> = ({
  schema,
  uiSchema,
  ...props
}) => {
  const {
    values: root,
    setValues: setRootValues,
  } = useFormikContext<ServiceSlice>();
  const [{ value }, , helpers] = useField<Expression<ServiceKeyVar>>(props);

  // This currently happens when a brick is copy-pasted into a separate extension
  // that does not yet have root.services configured, but already has the service
  // key set up in the (copied) BlockConfig.
  const isBadValue =
    value &&
    !root.services.some((service) =>
      isEqual(keyToFieldValue(service.outputKey), value)
    );

  useEffect(
    () => {
      if (value == null) {
        const match = root.services.find(
          (service) => service.id === PIXIEBRIX_SERVICE_ID
        );
        if (match?.outputKey) {
          // If the service is already being used, default to the currently configured auth
          console.debug(
            "PixieBrix dependency already exists for %s, using output key %s",
            match.id,
            match.outputKey,
            { root, match }
          );
          helpers.setValue(keyToFieldValue(match.outputKey));
        } else {
          console.debug("Adding PixieBrix API dependency");
          // Try defaulting to the only option available. Use onChange instead of helpers.setValue b/c it automatically
          // updates the services part of the form state
          setRootValues(
            produce(root, (draft) => {
              draft.services.push({
                id: PIXIEBRIX_SERVICE_ID,
                // XXX: in practice it the pixiebrix outputKey won't be used by any other service. However, we might
                // consider using fresh identifier here to eliminate the possibility of colliding with a different
                // service that is somehow already using the @pixiebrix key
                outputKey: PIXIEBRIX_OUTPUT_KEY,
                // PixieBrix service does not use an ID -- the auth is automatically handled based on the logged in user
                config: null,
              });
            })
          );
          helpers.setValue(keyToFieldValue(PIXIEBRIX_OUTPUT_KEY));
        }
      } else if (isBadValue) {
        // Clearing this "bad value" value will enable the preceding if-branch to
        // execute again, and that will configure root.services properly
        helpers.setValue(null);
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps -- run on mount, or if we detect a "bad value" (see comment above)
    [isBadValue]
  );

  return null;
};

export default AppServiceField;
