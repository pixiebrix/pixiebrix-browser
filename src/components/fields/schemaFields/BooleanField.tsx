/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import { fieldLabel } from "@/components/fields/fieldUtils";
import ConnectedFieldTemplate from "@/components/form/ConnectedFieldTemplate";
import SwitchButtonWidget from "@/components/form/widgets/switchButton/SwitchButtonWidget";

const BooleanField: React.FunctionComponent<SchemaFieldProps<boolean>> = ({
  name,
  label,
  schema,
}) => (
  <ConnectedFieldTemplate
    name={name}
    as={SwitchButtonWidget} // Do not use layout="switch" by default
    label={label ?? fieldLabel(name)}
    description={schema.description}
  />
);

export default BooleanField;
