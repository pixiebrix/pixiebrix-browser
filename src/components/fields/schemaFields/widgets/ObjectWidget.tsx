/* eslint-disable security/detect-object-injection -- working with object props a lot here */
/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useCallback, useMemo, useRef } from "react";
import { Button, Form, Table } from "react-bootstrap";
import { type SafeString, type Schema } from "@/core";
import { type SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import { isEmpty } from "lodash";
import { type FieldValidator, useField, useFormikContext } from "formik";
import { produce } from "immer";
import { freshIdentifier, joinName } from "@/utils";
import SchemaField from "@/components/fields/schemaFields/SchemaField";
import { getFieldNamesFromPathString } from "@/runtime/pathHelpers";
import { type UnknownObject } from "@/types";

type PropertyRowProps = {
  name: string;
  readOnly: boolean;
  schema: Schema;
  onDelete?: () => void;
  onRename?: (newName: string) => void;
  isRequired?: boolean;
  validate?: FieldValidator;
};

type RowProps = {
  parentSchema: Schema;
  name: string;
  property: string;
  defined: boolean;
  onDelete: (prop: string) => void;
  onRename: (oldProp: string, newProp: string) => void;
};

const CompositePropertyRow: React.FunctionComponent<PropertyRowProps> = ({
  name,
  schema,
  isRequired,
}) => (
  <tr>
    <td colSpan={2}>
      <SchemaField
        hideLabel
        isObjectProperty
        isRequired={isRequired}
        name={name}
        schema={schema}
      />
    </td>
  </tr>
);

const ValuePropertyRow: React.FunctionComponent<PropertyRowProps> = ({
  readOnly,
  onRename,
  schema,
  isRequired,
  name,
}) => {
  const [field] = useField(name);

  const updateName = useCallback(
    ({ target }: React.FocusEvent<HTMLInputElement>) => {
      onRename?.(target.value);
    },
    [onRename]
  );

  const currentProperty = getFieldNamesFromPathString(field.name)[1];

  return (
    <tr>
      <td>
        <Form.Control
          type="text"
          readOnly={readOnly}
          defaultValue={currentProperty}
          onBlur={updateName}
          data-testid={`${name}-name`}
        />
      </td>
      <td>
        <SchemaField
          hideLabel
          isObjectProperty
          {...field}
          schema={schema}
          isRequired={isRequired}
        />
      </td>
    </tr>
  );
};

const ObjectFieldRow: React.FunctionComponent<RowProps> = ({
  parentSchema,
  defined,
  name,
  property,
  onDelete,
  onRename,
}) => {
  const propertySchema: Schema = useMemo(() => {
    // As of v3, we allow object props of any type, not just string
    const defaultSchema: Schema = {};
    const rawSchema = defined
      ? parentSchema.properties[property]
      : parentSchema.additionalProperties ?? defaultSchema;

    return typeof rawSchema === "boolean" ? defaultSchema : rawSchema;
  }, [
    defined,
    parentSchema.properties,
    parentSchema.additionalProperties,
    property,
  ]);

  const isRequired = parentSchema.required?.includes(property) ?? false;

  const PropertyRowComponent = useMemo(
    () => getPropertyRow(propertySchema),
    [propertySchema]
  );

  const deleteProp = useCallback(() => {
    onDelete(property);
  }, [property, onDelete]);

  const renameProp = useCallback(
    (newProp: string) => {
      onRename(property, newProp);
    },
    [property, onRename]
  );

  return (
    <PropertyRowComponent
      key={property}
      name={name}
      readOnly={defined}
      schema={propertySchema}
      onDelete={defined ? undefined : deleteProp}
      onRename={defined ? undefined : renameProp}
      isRequired={isRequired}
    />
  );
};

export function getPropertyRow(
  schema: Schema
): React.FunctionComponent<PropertyRowProps> {
  switch (schema?.type) {
    case "array":
    case "object": {
      return CompositePropertyRow;
    }

    default: {
      return ValuePropertyRow;
    }
  }
}

type ObjectValue = UnknownObject;

const ObjectWidget: React.VFC<SchemaFieldProps> = (props) => {
  const { name, schema } = props;

  // Allow additional properties for empty schema (empty schema allows shape)
  const additionalProperties = isEmpty(schema) || schema.additionalProperties;

  // Helpers.setValue changes on every render, so use setFieldValue instead
  // https://github.com/formium/formik/issues/2268
  const [field] = useField<ObjectValue>(props);
  const { setFieldValue } = useFormikContext();

  // UseRef indirection layer so the callbacks below don't re-calculate on every change
  const valueRef = useRef(field.value);
  valueRef.current = field.value ?? {};

  const [properties, declaredProperties] = useMemo(() => {
    const declared = schema.properties ?? {};
    const additional = Object.fromEntries(
      Object.entries(field.value ?? {}).filter(
        ([property]) => !declared[property]
      )
    );
    return [[...Object.keys(declared), ...Object.keys(additional)], declared];
  }, [field.value, schema.properties]);

  const onDelete = useCallback(
    (property: string) => {
      setFieldValue(
        name,
        produce(valueRef.current, (draft) => {
          if (draft) {
            delete draft[property];
          }
        })
      );
    },
    [name, setFieldValue, valueRef]
  );

  const onRename = useCallback(
    (oldProp: string, newProp: string) => {
      if (oldProp !== newProp) {
        const previousValue = valueRef.current;

        console.debug("Renaming property", {
          newProp,
          oldProp,
          previousValue,
        });

        setFieldValue(
          name,
          produce(previousValue, (draft) => {
            draft[newProp] = draft[oldProp] ?? "";
            delete draft[oldProp];
          })
        );
      }
    },
    [name, setFieldValue, valueRef]
  );

  const addProperty = useCallback(() => {
    setFieldValue(
      name,
      produce(valueRef.current, (draft) => {
        const prop = freshIdentifier(
          "property" as SafeString,
          Object.keys(draft)
        );
        draft[prop] = "";
      })
    );
  }, [name, setFieldValue, valueRef]);

  return (
    <div>
      <Table size="sm" className="mb-0">
        <thead>
          <tr>
            <th scope="col">Property</th>
            <th scope="col">Value</th>
          </tr>
        </thead>
        <tbody>
          {properties.map((property) => (
            <ObjectFieldRow
              key={property}
              parentSchema={schema}
              name={
                // Always use nesting even if property name is empty
                property == null || property === ""
                  ? `${field.name}.`
                  : joinName(field.name, property)
              }
              property={property}
              defined={Object.hasOwn(declaredProperties, property)}
              onDelete={onDelete}
              onRename={onRename}
            />
          ))}
        </tbody>
      </Table>
      {additionalProperties && (
        <Button onClick={addProperty}>Add Property</Button>
      )}
    </div>
  );
};

export default ObjectWidget;
