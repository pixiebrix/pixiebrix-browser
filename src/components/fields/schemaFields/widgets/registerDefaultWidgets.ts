/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import UrlMatchPatternWidget from "./UrlMatchPatternWidget";
import UrlPatternWidget from "@/pageEditor/components/UrlPatternWidget";
import ArrayWidget from "./ArrayWidget";
import BooleanWidget from "./BooleanWidget";
import IntegerWidget from "./IntegerWidget";
import NumberWidget from "./NumberWidget";
import ObjectWidget from "./ObjectWidget";
import OmitFieldWidget from "./OmitFieldWidget";
import SchemaSelectWidget from "./SchemaSelectWidget";
import TemplateToggleWidget from "./TemplateToggleWidget";
import TextWidget from "./TextWidget";
import UnsupportedWidget from "./UnsupportedWidget";
import widgetsRegistry from "./widgetsRegistry";
import WorkshopMessageWidget from "./WorkshopMessageWidget";
import SelectorMatchWidget from "@/pageEditor/components/SelectorMatchWidget";
import CssClassWidget from "./CssClassWidget";
import DatabaseWidget from "./DatabaseWidget";
import PasswordWidget from "./PasswordWidget";
import IconWidget from "@/components/fields/IconWidget";

const defaultWidgets = {
  ArrayWidget,
  BooleanWidget,
  IntegerWidget,
  NumberWidget,
  ObjectWidget,
  OmitFieldWidget,
  SchemaSelectWidget,
  TemplateToggleWidget,
  TextWidget,
  UnsupportedWidget,
  UrlMatchPatternWidget,
  UrlPatternWidget,
  SelectorMatchWidget,
  CssClassWidget,
  WorkshopMessageWidget,
  DatabaseWidget,
  IconWidget,
  PasswordWidget,
} as const;

function registerDefaultWidgets() {
  Object.assign(widgetsRegistry, defaultWidgets);
  for (const [name, widget] of Object.entries(defaultWidgets)) {
    if (!widget) {
      throw new Error(
        `Error registering default widgets. ${name} is undefined. Is there a circular dependency?`
      );
    }
  }
}

export default registerDefaultWidgets;
