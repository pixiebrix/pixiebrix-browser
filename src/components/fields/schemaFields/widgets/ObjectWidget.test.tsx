/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { Schema } from "@/core";
import { render, screen } from "@/testUtils/testHelpers";
import ObjectWidget from "@/components/fields/schemaFields/widgets/ObjectWidget";
import userEvent from "@testing-library/user-event";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import { stringToExpression } from "@/pageEditor/extensionPoints/upgrade";
import { expectToggleOptions } from "@/components/fields/schemaFields/fieldTestUtils";

const fieldName = "testField";
const fieldDescription = "this is a test field description";

describe("ObjectWidget", () => {
  beforeAll(() => {
    registerDefaultWidgets();
  });

  test("renders empty widget", () => {
    const schema: Schema = {
      type: "object",
      additionalProperties: true,
    };
    expect(
      render(<ObjectWidget name={fieldName} schema={schema} />, {
        initialValues: {
          [fieldName]: {},
        },
      }).asFragment()
    ).toMatchSnapshot();
  });

  test("renders widget with properties", () => {
    const schema: Schema = {
      type: "object",
      properties: {
        foo: {
          type: "number",
          default: 42,
        },
        bar: {
          type: "string",
        },
        baz: {
          oneOf: [
            {
              type: "boolean",
            },
            {
              type: "number",
            },
            {
              type: "string",
            },
          ],
        },
      },
    };
    expect(
      render(<ObjectWidget name={fieldName} schema={schema} />, {
        initialValues: {
          [fieldName]: {},
        },
      }).asFragment()
    ).toMatchSnapshot();
  });

  test("add property and change value", async () => {
    const schema: Schema = {
      type: "object",
      additionalProperties: true,
    };
    const { getFormState } = render(
      <ObjectWidget
        name={fieldName}
        schema={schema}
        isRequired
        description={fieldDescription}
        defaultType="object"
      />,
      {
        initialValues: {
          [fieldName]: {},
        },
      }
    );

    // Add a property
    await userEvent.click(screen.getByText(/add property/i));

    const [nameInput, valueInput] = screen.getAllByRole("textbox");

    // Change the property name
    await userEvent.clear(nameInput);
    await userEvent.type(nameInput, "myProp");

    // Change the value
    await userEvent.click(valueInput);
    await userEvent.type(screen.getAllByRole("textbox")[1], "myValue");

    // Blur the value input to set the value
    await userEvent.click(nameInput);

    const formState = await getFormState();

    expect(formState).toStrictEqual({
      [fieldName]: {
        myProp: stringToExpression("myValue", "nunjucks"),
      },
    });
  });

  test("remove property", async () => {
    const schema: Schema = {
      type: "object",
      additionalProperties: {
        type: "string",
      },
    };
    const { getFormState } = render(
      <ObjectWidget
        name={fieldName}
        schema={schema}
        isRequired
        description={fieldDescription}
        defaultType="object"
      />,
      {
        initialValues: {
          [fieldName]: {
            foo: "bar",
            myField: "myValue",
          },
        },
      }
    );

    // Open the field type toggle
    await userEvent.click(
      screen.getByTestId(`toggle-${fieldName}.myField`).querySelector("button")
    );

    // Select "Exclude"
    await userEvent.click(screen.getByText("Exclude"));

    const formState = await getFormState();

    // Expect excluded property to be removed from the state
    expect(formState).toStrictEqual({
      [fieldName]: {
        foo: "bar",
      },
    });
  });

  test("defined property removed from state and not ui", async () => {
    const schema: Schema = {
      type: "object",
      properties: {
        foo: {
          type: "string",
        },
        bar: {
          type: "string",
        },
      },
    };
    const { getFormState } = render(
      <ObjectWidget
        name={fieldName}
        schema={schema}
        isRequired
        description={fieldDescription}
        defaultType="object"
      />,
      {
        initialValues: {
          [fieldName]: {
            foo: "fooValue",
            bar: "barValue",
          },
        },
      }
    );

    // Open the field type toggle
    await userEvent.click(
      screen.getByTestId(`toggle-${fieldName}.bar`).querySelector("button")
    );

    // Select "Exclude"
    await userEvent.click(screen.getByText("Exclude"));

    const formState = await getFormState();

    // Expect excluded property to be removed from the state
    expect(formState).toStrictEqual({
      [fieldName]: {
        foo: "fooValue",
      },
    });

    // Expect defined property that was "removed" to remain in the ui
    expect(screen.getByDisplayValue("bar")).toBeVisible();
  });

  test("required property can't be removed", async () => {
    const schema: Schema = {
      type: "object",
      properties: {
        foo: {
          type: "string",
        },
        bar: {
          type: "string",
        },
      },
      required: ["foo"],
    };
    const { container } = render(
      <ObjectWidget
        name={fieldName}
        schema={schema}
        isRequired
        description={fieldDescription}
        defaultType="object"
      />,
      {
        initialValues: {
          [fieldName]: {
            foo: "fooValue",
            bar: "barValue",
          },
        },
      }
    );

    // Find the field type toggles
    const fooToggle = screen
      .getByTestId(`toggle-${fieldName}.foo`)
      .querySelector("button");
    const barToggle = screen
      .getByTestId(`toggle-${fieldName}.bar`)
      .querySelector("button");

    // Open foo's field type toggle
    await userEvent.click(fooToggle);

    // Expect not to find "Exclude" for a required field
    await expectToggleOptions(container, ["string", "var"]);

    // Close foo's toggle
    await userEvent.click(fooToggle);

    // Open bar's field type toggle
    await userEvent.click(barToggle);

    // Expect to find the "Exclude" (omit) option
    await expectToggleOptions(container, ["string", "var", "omit"]);
  });
});
