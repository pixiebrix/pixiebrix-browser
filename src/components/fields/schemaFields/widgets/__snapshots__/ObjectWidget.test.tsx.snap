// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectWidget renders empty widget 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div>
      <table
        class="mb-0 table table-sm"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              Property
            </th>
            <th
              scope="col"
            >
              Value
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
      <button
        class="btn btn-primary"
        type="button"
      >
        Add Property
      </button>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`ObjectWidget renders widget with properties 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div>
      <table
        class="mb-0 table table-sm"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              Property
            </th>
            <th
              scope="col"
            >
              Value
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <input
                class="form-control"
                data-testid="testField.foo-name"
                readonly=""
                type="text"
                value="foo"
              />
            </td>
            <td>
              <div
                class="formGroup mb-0 form-group row"
              >
                <div
                  class="col-xl-12 col-lg-12"
                >
                  <div
                    class="root"
                  >
                    <div
                      class="field"
                    >
                      <div
                        role="button"
                        tabindex="0"
                      >
                        <input
                          class="form-control"
                          disabled=""
                          id="testField.foo"
                          name="testField.foo"
                          value=""
                        />
                      </div>
                    </div>
                    <div
                      class="dropdown dropdown"
                      data-test-selected="Exclude"
                      data-testid="toggle-testField.foo"
                    >
                      <button
                        aria-expanded="false"
                        aria-haspopup="true"
                        class="dropdown-toggle btn btn-link"
                        type="button"
                      >
                        <span
                          class="symbol"
                        >
                          <test-file-stub
                            classname="root"
                          />
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </td>
          </tr>
          <tr>
            <td>
              <input
                class="form-control"
                data-testid="testField.bar-name"
                readonly=""
                type="text"
                value="bar"
              />
            </td>
            <td>
              <div
                class="formGroup mb-0 form-group row"
              >
                <div
                  class="col-xl-12 col-lg-12"
                >
                  <div
                    class="root"
                  >
                    <div
                      class="field"
                    >
                      <div
                        role="button"
                        tabindex="0"
                      >
                        <input
                          class="form-control"
                          disabled=""
                          id="testField.bar"
                          name="testField.bar"
                          value=""
                        />
                      </div>
                    </div>
                    <div
                      class="dropdown dropdown"
                      data-test-selected="Exclude"
                      data-testid="toggle-testField.bar"
                    >
                      <button
                        aria-expanded="false"
                        aria-haspopup="true"
                        class="dropdown-toggle btn btn-link"
                        type="button"
                      >
                        <span
                          class="symbol"
                        >
                          <test-file-stub
                            classname="root"
                          />
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </td>
          </tr>
          <tr>
            <td>
              <input
                class="form-control"
                data-testid="testField.baz-name"
                readonly=""
                type="text"
                value="baz"
              />
            </td>
            <td>
              <div
                class="formGroup mb-0 form-group row"
              >
                <div
                  class="col-xl-12 col-lg-12"
                >
                  <div
                    class="root"
                  >
                    <div
                      class="field"
                    >
                      <div
                        role="button"
                        tabindex="0"
                      >
                        <input
                          class="form-control"
                          disabled=""
                          id="testField.baz"
                          name="testField.baz"
                          value=""
                        />
                      </div>
                    </div>
                    <div
                      class="dropdown dropdown"
                      data-test-selected="Exclude"
                      data-testid="toggle-testField.baz"
                    >
                      <button
                        aria-expanded="false"
                        aria-haspopup="true"
                        class="dropdown-toggle btn btn-link"
                        type="button"
                      >
                        <span
                          class="symbol"
                        >
                          <test-file-stub
                            classname="root"
                          />
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
